pkg/gencosmosdb/cosmosdb/template_fake.go

```diff
@@ -0,0 +1,301 @@
+package cosmosdb
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/ugorji/go/codec"
+
+	pkg "github.com/jim-minter/go-cosmosdb/pkg/gencosmosdb/cosmosdb/dummy"
+)
+
+type FakeTemplateDecoder func(*string) *pkg.Template
+type FakeTemplateTrigger func(context.Context, *pkg.Template) error
+type FakeTemplateQuery func(*Query, map[string]*string, FakeTemplateDecoder) []*string
```

Comment:
The decoder being passed around is because `SecureString`'s encoders in the JSONHandle :( I guess we can just make a new one in the client though, since the client is the one that makes it to begin with...

Reply:
Okay, I've gone over this again and I've settled on `type FakeTemplateQuery func(TemplateClient, *Query) TemplateRawIterator` as a shape that seems reasonable. I've also moved to a RWMutex, which lets us lock when running the query in a way that lets the Query itself make calls. I've also shifted the code in ARO-RP to do so as well, which means that we don't need to pass around the raw mapping, or the JSONHandle at all.

I've then added an implementation for the one special case where TemplateRawIterator was more complex than it needed to be (since it did a query that didn't return documents, but arbitrary JSON) in ARO-RP, and added NewFakeTemplateClientRawIterator for when we simply want to boilerplate out returning a list of documents, and I think is worth keeping as canned functionality. This simplifies things a lot.